### PR TITLE
feat(binding-mcp): implement mcp · server binding

### LINKED ISSUE #1668: binding-mcp: implement mcp · server binding
## Summary

Implement the `mcp · server` binding in a new `binding-mcp` module. Accepts `http`, `ws`, or `sse` streams from AI agents and produces `mcp` streams into the binding pipeline.

## Stream type

- Accepts: `http` / `ws` / `sse`
- Produces: `mcp`

## Responsibilities

- **Transport**: SSE, HTTP, and WebSocket — all three MCP transports
- **Session lifecycle**: initialize handshake, capability negotiation
- **Local prompts (additive)**: locally declared prompts are merged with federated upstream results; for `prompts/get`, local prompt names are served directly and toolkit-prefixed names are routed downstream
- **Callback stream**: OAuth authorization code callbacks from browser are passed downstream unchanged as a `callback` mcp stream type
- **Guard integration**: JWT guard validation of inbound bearer tokens on routes
- **URL elicitation**: returns a JSON-RPC `resource_required` error to the agent when downstream signals resource_required, embedding the authorization URL
- Path routing is handled upstream by the `http · server` binding — not redeclared here

## MCP stream sub-types

| Sub-type | Trigger | Description |
|----------|---------|-------------|
| `tool` | `tools/call` | Tool invocation |
| `resource` | `resources/read` | Resource read |
| `prompt` | `prompts/get` | Prompt retrieval |
| `callback` | HTTP GET from browser | OAuth authorization code callback — passed downstream unchanged |

## Configuration

```yaml
mcp_server0:
  type: mcp
  kind: server
  routes:
    - guarded:
        jwt0:
          - mcp:access
      exit: mcp_proxy0
```

### Locally declared prompts

```yaml
mcp_server0:
  type: mcp
  kind: server
  options:
    prompts:
      review_pr:
        description: >
          Structured prompt for reviewing a pull request for
          correctness, test coverage, and edge cases.
        arguments:
          model: json
          catalog:
            prompts_catalog0:
              subject: review_pr_args
        template:
          - role: user
            content: >
              Please review PR #${args.pr_number} in
              ${args.owner}/${args.repo}. Focus on correctness,
              test coverage, and potential edge cases.
      summarise_diff:
        description: >
          Few-shot prompt for summarising a PR diff.
        arguments:
          model: json
          catalog:
            prompts_catalog0:
              subject: review_pr_args
        template:
          - role: user
            content: Summarise PR #1 in aklivity/zilla
          - role: assistant
            content: >
              This PR adds MCP gateway support. Key changes:
              new mcp server, proxy, and client bindings.
          - role: user
            content: >
              Now summarise PR #${args.pr_number} in
              ${args.owner}/${args.repo}.
```

## Prompt template roles

| Role | Purpose |
|------|---------|
| `user` | The task instruction or request — the common case |
| `assistant` | Pre-filled response for few-shot prompting — guides model style, format, or reasoning pattern |

## Implementation Notes
This binding should first implement http-only, non streaming mapping from `http` network stream to `mcp` application stream.
Then, to support `sse` response, this binding should create a composite namespace with sse binding, and split the network `http` stream so that the request still maps to an application `mcp` stream, and also sets up a response sse stream that routes back to the same binding, but detected to come from the composite namespace and is joined with the `mcp` application stream so that responses are sent back to the client through the `sse` binding in the composite namespace.
Then to support `ws`, which is bidirectional, the `http` network stream is routed instead to a `ws` binding in the composite namespace, which then routes back to the `mcp` server binding, similar to `sse` approach, but for both directions, then the `mcp` application stream continues following the configured route exit.
The `openapi` binding shows how a composite binding works.
The `http-kafka` binding shows how a single `http` stream can map to separate streams for request and response.

## Additional Context
[MCP Specification](https://modelcontextprotocol.io/specification/2025-11-25)
[MCP URL Elicitation Required](https://modelcontextprotocol.io/specification/draft/client/elicitation#url-elicitation-required-error)
[MCP Streamable HTTP Transport](https://modelcontextprotocol.io/specification/2025-11-25/basic/transports#streamable-http)
[MCP Protocol Overview](https://modelcontextprotocol.io/specification/2025-11-25/basic)
[MCP Protocol Lifecycle](https://modelcontextprotocol.io/specification/2025-11-25/basic/lifecycle)
[MCP Protocol Cancelation](https://modelcontextprotocol.io/specification/2025-11-25/basic/utilities/cancellation)
[MCP Protocol Ping](https://modelcontextprotocol.io/specification/2025-11-25/basic/utilities/ping)
[MCP Protocol Progress](https://modelcontextprotocol.io/specification/2025-11-25/basic/utilities/progress)

### PR DESCRIPTION
**Pull request by jfallows**
## Summary

- Implements the foundational `runtime/binding-mcp` module with `server` kind handler for the MCP (Model Context Protocol) binding
- Adds `McpBeginEx` IDL type defining `kind` and `sessionId` fields for MCP stream extension
- Registers `McpBindingFactorySpi`, `McpOptionsConfigAdapter`, and `McpConditionConfigAdapter` via SPI and `module-info.java`
- Registers `binding-mcp` and `binding-mcp.spec` in `runtime/pom.xml` and `specs/pom.xml`

## Test plan

- [ ] Verify `./mvnw install -DskipTests -pl runtime/binding-mcp -am` passes
- [ ] Verify `McpBindingFactorySpi` registers type `"mcp"` correctly
- [ ] Verify schema patch `mcp.schema.patch.json` validates `type: mcp, kind: server` configs
- [ ] Verify `McpOptionsConfigAdapter` round-trips `prompts` list correctly
- [ ] Verify `McpConditionConfigAdapter` round-trips `kind` field correctly

References #1668

https://claude.ai/code/session_0174raBeXFTgt98bp4DTyRDm

### CODE REVIEW

File: runtime/engine/src/main/java/io/aklivity/zilla/runtime/engine/EngineConfiguration.java [hunk 1]
```diff
@@ -553,6 +560,15 @@ private static String defaultName(
         return System.getProperty(ZILLA_NAME_PROPERTY, "zilla");
     }
 
+    private static String defaultVersion(
+        Configuration config)
+    {
+        java.lang.module.ModuleDescriptor descriptor = EngineConfiguration.class.getModule().getDescriptor();
+        return descriptor != null
+            ? descriptor.version().map(java.lang.module.ModuleDescriptor.Version::toString).orElse("develop-SNAPSHOT")
+            : "develop-SNAPSHOT";
```
Comment (jfallows) [inline runtime/engine/src/main/java/io/aklivity/zilla/runtime/engine/EngineConfiguration.java:577]: Use import for module descriptor type, not fully qualified class name.

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpServerFactory.java [hunk 2]
```diff
@@ -0,0 +1,2795 @@
...
+                break;
+            case "prompts/get":
+                server.onDecodePromptsGet(value, traceId, authorization);
+                break;
+            case "resources/read":
+                server.onDecodeResourcesRead(value, traceId, authorization);
+                break;
+            }
+
+            parser.skipObject(); // TODO: non-blocking
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpServerFactory.java]: `parser.skipObject()` is blocking, but we may not have enough input yet, so we need to change this to non-blocking by introducing a decodeJsonRpcSkipObject method and a way to indicate which decoder to use when the object is skipped.

Note that `skipObject()` assumes already inside the object, so initial depth is intially 1 (not zero). Nested `START_OBJECT` increase the depth, while `END_OBJECT` decreases the depth. When the depth gets back to zero, then transition to the next decoder state, specified in a server field such as `decodeAfterSkipObject`, while the actual decoder state is set to `decodeJsonRpcSkipObject`.

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpServerFactory.java [hunk 3]
```diff
@@ -622,8 +623,9 @@ private int decodeJsonRpcParamsStart(
             }
             else
             {
-                parser.skipObject(); // TODO: non-blocking
-                server.decoder = decodeJsonRpcParamsEnd;
+                server.skipObjectDepth = 1;
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpServerFactory.java]: Naming convention for all decode state fields is to name `decodedXXX`, so let's rename `server.skipObjectDepth` to `server.decodedSkipObjectDepth`

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpServerFactory.java [hunk 4]
```diff
@@ -692,8 +694,10 @@ private int decodeJsonRpcParamsSkipValue(
...
             {
             case JsonParser.Event.START_OBJECT:
-                parser.skipObject(); // TODO: non-blocking
-                break;
+                server.skipObjectDepth = 1;
+                server.decodeAfterSkipObject = decodeJsonRpcMethodWithParam;
+                server.decoder = decodeJsonRpcSkipObject;
+                break decode;
             case JsonParser.Event.START_ARRAY:
                 parser.skipArray(); // TODO: non-blocking
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpServerFactory.java]: Implement a similar non-blocking pattern to replace `parser.skipArray()`.

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpServerFactory.java [hunk 5]
```diff
@@ -839,6 +880,8 @@ private final class McpServer
         private long encodeSlotTraceId;
 
         private McpServerDecoder decoder;
+        private McpServerDecoder decodeAfterSkipObject;
+        private int skipObjectDepth;
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpServerFactory.java]: move these fields down into the next block with other decode state.

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpServerFactory.java [hunk 6]
```diff
@@ -623,7 +624,7 @@ private int decodeJsonRpcParamsStart(
             }
             else
             {
-                server.skipObjectDepth = 1;
+                server.decodedSkipObjectDepth = 1;
                 server.decodeAfterSkipObject = decodeJsonRpcParamsEnd;
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpServerFactory.java]: Rename `server.decodeAfterSkipObject` to `server.decodedSkipObjectThen`

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpServerFactory.java [hunk 7]
```diff
@@ -694,20 +695,20 @@ private int decodeJsonRpcParamsSkipValue(
...
-                server.skipObjectDepth = 1;
+                server.decodedSkipObjectDepth = 1;
                 server.decodeAfterSkipObject = decodeJsonRpcMethodWithParam;
                 server.decoder = decodeJsonRpcSkipObject;
                 break decode;
             case JsonParser.Event.START_ARRAY:
-                parser.skipArray(); // TODO: non-blocking
-                break;
+                server.decodedSkipArrayDepth = 1;
+                server.decodeAfterSkipArray = decodeJsonRpcMethodWithParam;
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpServerFactory.java]: Rename to `decodedSkipArrayThen`

**Review by jfallows — COMMENTED** [review]
Looks good.

**Review by jfallows — COMMENTED** [review]
LGTM

File: runtime/binding-mcp/src/test/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpServerIT.java [hunk 8]
```diff
@@ -87,11 +90,24 @@ public void shouldShutdownLifecycle() throws Exception
...
         k3po.finish();
     }
 
+    @Test
+    @Configuration("server.yaml")
+    @Specification({
+        "${net}/lifecycle.inactivity.timeout/client",
+        "${app}/lifecycle.inactivity.timeout/server"})
+    @Configure(name = MCP_INACTIVITY_TIMEOUT_NAME, value = "PT1S")
+    public void shouldTimeoutInactiveSession() throws Exception
```
Comment (jfallows) [inline runtime/binding-mcp/src/test/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpServerIT.java]:
```suggestion
    @Test
    @Configuration("server.yaml")
    @Specification({
        "${net}/lifecycle.timeout/client",
        "${app}/lifecycle.timeout/server"})
    @Configure(name = MCP_INACTIVITY_TIMEOUT_NAME, value = "PT1S")
    public void shouldTimeoutLifecycle() throws Exception
```

File: specs/binding-mcp.spec/src/main/scripts/io/aklivity/zilla/specs/binding/mcp/streams/network/lifecycle.timeout/client.rpt
Comment (jfallows) [inline specs/binding-mcp.spec/src/main/scripts/io/aklivity/zilla/specs/binding/mcp/streams/network/lifecycle.timeout/client.rpt:1]: rename scenario to `lifecycle.timeout`

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpServerFactory.java [hunk 9]
```diff
@@ -0,0 +1,2985 @@
...
+        private void touch()
+        {
+            lastActiveAt = System.currentTimeMillis();
+        }
+
+        private void scheduleInactivity(
+            long traceId)
+        {
+            final long at = lastActiveAt + inactivityTimeoutMillis;
+            inactivityTimeoutId = signaler.signalAt(at, originId, routedId, initialId,
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpServerFactory.java]:
```suggestion
            inactiveId = signaler.signalAt(at, originId, routedId, initialId,
```

File: runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpServerFactory.java [hunk 10]
```diff
@@ -0,0 +1,2985 @@
...
+                traceId, INACTIVITY_SIGNAL_ID, 0);
```
Comment (jfallows) [inline runtime/binding-mcp/src/main/java/io/aklivity/zilla/runtime/binding/mcp/internal/stream/McpServerFactory.java]:
```suggestion
                traceId, INACTIVE_SIGNAL_ID, 0);
```

**Review by jfallows — COMMENTED** [review]
LGTM
